src/rust/engine/src/externs/interface.rs

```diff
@@ -1671,16 +1672,26 @@ fn write_digest(
     }
 
     block_in_place_and_wait(py, || async move {
-      core
-        .store()
+      let store = core.store();
+      store
         .materialize_directory(
           destination.clone(),
-          lifted_digest,
+          lifted_digest.clone(),
           true, // Force everything we write to be mutable
           &BTreeSet::new(),
           fs::Permissions::Writable,
         )
-        .await
+        .await?;
+
+      if invalidate {
+        let snapshot = store::Snapshot::from_digest(store, lifted_digest).await?;
+        let files = snapshot.tree.files(SymlinkBehavior::Aware);
```

Comment:
It looks like the digest and thus the snapshot/`files` here contains _all_ files for goals like `pants fmt ::` (or `fix`), even unchanged ones, so this spuriously invalidates them too. Is that a problem? Should we be doing something with a snapshot diff in those callers?

Reply:
It shouldn't be a problem, because we were already spuriously invalidating them by overwriting them (`materialize_directory` never checks the content of the thing it is overwriting).

The `graph` implements early-cutoff, meaning that if an invalidated node hasn't actually changed, it is "cleaned" in-place without its dependents re-running.

Reply:
Also, you should technically include empty directories and symlinks here as well: basically, all paths in the `DigestTrie` in a symlink-oblivious way. We don't have a method for that currently, but you could add it as something like:
```rust
  /// The paths of all "leaf" nodes of the DigestTrie: empty directories, files, or symlinks.
  pub fn leaf_paths(&self) -> Vec<PathBuf> {
    let mut paths = Vec::new();
    self.walk(symlink_behavior, &mut |path, entry| {
      match entry {
        Entry::Directory(d) if d.tree.is_empty() => paths.push(path.to_owned()),
        Entry::File(_) | Entry::Symlink(_) => paths.push(path.to_owned())
        _ => (),
      }
    });
    paths
  }
```